File: 2024/matlab/aoc12.m
clear; close all;
input_path = fullfile(get_input_root,'day_12.txt');
x = strip(read_txt(input_path));
lines = strsplit(x,crlf);
X = double(cat(1,lines{:}));

ux = unique(X(:));
X1 = zeros(size(X,1)+2,size(X,2)+2);
X1(2:end-1,2:end-1) = X;
X = X1;

k = [0,1,0;1,1,1;0,1,0];
%part 1 %%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
P = 0;
for i1 = 1:numel(ux)
    Xi = ux(i1) == X;
    XiL = bwlabel(Xi,4);
    uxi = unique(XiL);
    uxi = uxi(2:end);
    for i2 = 1:numel(uxi)
        Xi2 = uxi(i2) == XiL;
        Xe = imdilate(Xi2,k);
        Xes = conv2(Xi2,k,'same');
        Xp = Xe-Xi2;
        Xp = Xp.*Xes;
        P = P + sum(Xp(:))*sum(Xi2(:));
    end
end
disp(P)

%part 2 %%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
ksb = [1,0;0,0];
kd = {[1,0;0,1];[0,1;1,0]};
ks = [];
for i1 = 1:4
    ks = cat(1,ks,{rot90(ksb,i1)});
    ks = cat(1,ks,{rot90(~ksb,i1)});
end
P = 0;
for i1 = 1:numel(ux)
    Xi = ux(i1) == X;
    XiL = bwlabel(Xi,4);
    uxi = unique(XiL);
    uxi = uxi(2:end);
    for i2 = 1:numel(uxi)
        Xi2 = uxi(i2) == XiL;
        sides = 0;
        for i3 = 1:numel(kd)
            i_match = abs(normxcorr2(kd{i3},Xi2) - 1)<10*eps;
            sides = sides + 2*sum(i_match(:));
        end
        for i3 = 1:numel(ks)
            i_match = abs(normxcorr2(ks{i3},Xi2) - 1)<10*eps;
            sides = sides + 1*sum(i_match(:));
        end
        P = P + sides*sum(Xi2(:));
    end
end
disp(P)

File: 2024/python/aoc12.py
# -*- coding: utf-8 -*-
import numpy as np
from scipy.signal import convolve2d as conv2
from scipy.ndimage import label

#prepare variables ############################################################
input_path = '..\inputs\day_12.txt'
lines = open(input_path).read().split('\n')
X = np.zeros((len(lines)+2,len(lines[0])+2))
for i,line in enumerate(lines):
    X[i+1,1:-1] = np.array([ord(c) for c in line]) #numeric into padded array
    
#part 1 #######################################################################    
k = np.array([[0,1,0],[1,1,1],[0,1,0]])
uX = np.unique(X.flatten())[1:]
P = 0
for v in uX:
    XL = label(v==X,k)[0]
    uXL = np.unique(XL.flatten())[1:]
    for u in uXL:
        Xum = (u==XL).astype(int)
        Xe = conv2(X,k,'same')>0 #image dilation
        Xes = conv2(Xum,k,'same')
        Xp = (Xe-Xum)*Xes
        P += sum(Xp.flatten())*sum(Xum.flatten())
print(P)

#part 2 #######################################################################
def corr2(X,k):
    X = X - np.mean(X.flatten())
    k = k - np.mean(k.flatten())
    return conv2(X,k,'same')/np.sum(k.flatten()**2)

kd = [np.array([[1,0],[0,1]]),np.array([[0,1],[1,0]])]
ks = [];
ksb = np.array([[1,0],[0,0]])
for i1 in range(0, 4):  
    ks.extend([np.rot90(1-ksb, i1),np.rot90(ksb, i1)])         
P = 0;
for v in uX:
    XL = label(v==X,k)[0]
    uXL = np.unique(XL.flatten())[1:]
    for u in uXL:
        Xum = (u==XL).astype(int)
        sides = 0
        for kde in kd:
            sides = sides + 2*np.sum(np.isclose(corr2(Xum,kde),1))
        for kse in ks:
            sides = sides + 1*np.sum(np.isclose(corr2(Xum,kse),1))
        P = P + sides*sum(Xum.flatten())
print(P)